python/foxglove-sdk-examples/ws-param-server/main.py

```diff
@@ -0,0 +1,97 @@
+"""
+This implements a parameter server for live visualization.
+
+View and edit parameters from a Parameters panel in Foxglove:
+https://docs.foxglove.dev/docs/visualization/panels/parameters
+"""
+
+import logging
+from typing import Optional
+import foxglove
+import time
+
+
+from foxglove import Capability
+from foxglove import ParameterType, ParameterValue, Parameter
+
+
+class ParameterStore(foxglove.ServerListener):
+    def __init__(self, parameters: list[Parameter]) -> None:
+        self.parameters = parameters
+
+    # Foxglove server callback
+    def on_get_parameters(
+        self,
+        client: foxglove.Client,
+        param_names: list[str],
+        request_id: Optional[str] = None,
+    ) -> list[Parameter]:
+        logging.debug(f"on_get_parameters: {param_names}, {client.id}, {request_id}")
+        return self.parameters
+
+    def on_set_parameters(
+        self,
+        client: foxglove.Client,
+        parameters: list[foxglove.Parameter],
+        request_id: str | None = None,
+    ) -> list[foxglove.Parameter]:
+        logging.debug(f"on_set_parameters: {parameters}, {client.id}, {request_id}")
+        existing_names = [p.name for p in self.parameters]
+        for changed_param in parameters:
+            if changed_param.name not in existing_names:
+                # Add
+                self.parameters.append(changed_param)
+            elif changed_param.value is None:
+                # Remove
+                self.parameters = [
+                    param
+                    for param in self.parameters
+                    if param.name != changed_param.name
+                ]
+            else:
+                # Update
+                self.parameters = [
+                    param if param.name != changed_param.name else changed_param
+                    for param in self.parameters
+                ]
+        return parameters
+
+
+def main() -> None:
+    foxglove.verbose_on()
+
+    initial_values: list[Parameter] = [
+        Parameter(
+            "param1",
+            value=ParameterValue.Dict(
```

Comment:
I see what you mean. This would be a much nicer interface in Python if you could just use native lists, dicts, bytes, etc and we did the conversion in our code.

Although we'd have to throw an exception at runtime if you passed the wrong type.

I feel like we could provide the correct typings though. Also this current code will still throw an exception at runtime if you pass the wrong types. So it's only slightly safer.

I guess for now this works and we can make it more ergonomic in a future iteration.

Reply:
I think it's reasonable to support both, so I'll keep this and have also filed FG-10589 for the alternative.